shopinvader_address/models/res_partner.py

```diff
@@ -0,0 +1,177 @@
+# Copyright 2023 ACSONE SA/NV
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import _, models
+from odoo.exceptions import MissingError, UserError
+
+
+class ResPartner(models.Model):
+
+    _inherit = "res.partner"
+
+    def _ensure_shopinvader_billing_address_not_used(self) -> None:
+        """
+        Check if Billing Address is used on confirmed sale order
+        """
+        self.ensure_one()
+        sale_order = (
+            self.env["sale.order"]
+            .sudo()
+            .search(
+                [
+                    ("partner_invoice_id", "=", self.id),
+                    ("state", "in", ("done", "sale")),
+                ],
+                limit=1,
+            )
+        )
+        if len(sale_order) > 0:
+            raise UserError(
+                _(
+                    "Can not update billing addresses(%(address_id)d)"
+                    "because it is already used on confirmed sale order",
+                    address_id=self.id,
+                )
+            )
+
+    def _ensure_shopinvader_shipping_address_not_used(self) -> None:
+        """
+        Check if Shipping Address is used on confirmed sale order
+        """
+        self.ensure_one()
+        sale_order = (
+            self.env["sale.order"]
+            .sudo()
+            .search(
+                [
+                    ("partner_shipping_id", "=", self.id),
+                    ("state", "in", ("done", "sale")),
+                ],
+                limit=1,
+            )
+        )
+        if len(sale_order) > 0:
+            raise UserError(
+                _(
+                    "Can not delete Shipping address(%(address_id)d)"
+                    "because it is already used on confirmed sale order",
+                    address_id=self.id,
+                )
+            )
+
+    # --- Billing ---
+    # Billing addresses is unique and corresponds to authenticated_partner
+
+    def _get_shopinvader_billing_addresses(self) -> "ResPartner":
+        self.ensure_one()
+        return self
+
+    def _get_shopinvader_billing_address(self, address_id: int) -> "ResPartner":
+        self.ensure_one()
+        addresses = self._get_shopinvader_billing_addresses()
+        address = addresses.filtered(lambda rec: rec.id == address_id)
+        if not address:
+            raise MissingError(
+                _(
+                    "Billing address not found, id: %(address_id)d",
+                    address_id=address_id,
+                )
+            )
```

Comment:
Yes, it's a good idea to put this raise MissingError in this domain model function instead of the service. :+1: